File: native/android/VueNativeCore/src/main/kotlin/com/vuenative/core/Modules/AsyncStorageModule.kt
package com.vuenative.core

import android.content.Context
import android.content.SharedPreferences

class AsyncStorageModule : NativeModule {
    override val moduleName = "AsyncStorage"
    private var prefs: SharedPreferences? = null

    override fun initialize(context: Context, bridge: NativeBridge) {
        prefs = context.getSharedPreferences("vue_native_async_storage", Context.MODE_PRIVATE)
    }

    override fun invoke(method: String, args: List<Any?>, bridge: NativeBridge, callback: (Any?, String?) -> Unit) {
        val p = prefs ?: run { callback(null, "Storage not initialized"); return }
        when (method) {
            "getItem" -> {
                val key = args.getOrNull(0)?.toString() ?: run { callback(null, "Missing key"); return }
                callback(p.getString(key, null), null)
            }
            "setItem" -> {
                val key   = args.getOrNull(0)?.toString() ?: run { callback(null, "Missing key"); return }
                val value = args.getOrNull(1)?.toString() ?: run { callback(null, "Missing value"); return }
                p.edit().putString(key, value).apply()
                callback(null, null)
            }
            "removeItem" -> {
                val key = args.getOrNull(0)?.toString() ?: run { callback(null, "Missing key"); return }
                p.edit().remove(key).apply()
                callback(null, null)
            }
            "getAllKeys" -> {
                callback(p.all.keys.toList(), null)
            }
            "clear" -> {
                p.edit().clear().apply()
                callback(null, null)
            }
            "multiGet" -> {
                val keys = (args.getOrNull(0) as? List<*>)?.map { it?.toString() ?: "" } ?: emptyList()
                val result = keys.map { k -> listOf(k, p.getString(k, null)) }
                callback(result, null)
            }
            "multiSet" -> {
                val pairs = args.getOrNull(0) as? List<*> ?: emptyList<Any>()
                val editor = p.edit()
                pairs.forEach { pair ->
                    val kv = pair as? List<*>
                    val k = kv?.getOrNull(0)?.toString()
                    val v = kv?.getOrNull(1)?.toString()
                    if (k != null && v != null) editor.putString(k, v)
                }
                editor.apply()
                callback(null, null)
            }
            else -> callback(null, "Unknown method: $method")
        }
    }
}

File: native/shared/VueNativeShared/Sources/VueNativeShared/Modules/AsyncStorageModule.swift
import Foundation

/// Native module providing async key-value storage backed by UserDefaults.
///
/// Methods:
///   - getItem(key: String) -> String?
///   - setItem(key: String, value: String)
///   - removeItem(key: String)
///   - getAllKeys() -> [String]
///   - clear()
public final class AsyncStorageModule: NativeModule {
    public let moduleName = "AsyncStorage"

    private let defaults = UserDefaults.standard
    private let keyPrefix = "VueNative.AsyncStorage."

    public init() {}

    public func invoke(method: String, args: [Any], callback: @escaping (Any?, String?) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            switch method {
            case "getItem":
                guard let key = args.first as? String else {
                    callback(nil, "getItem: missing key")
                    return
                }
                let result = self.defaults.string(forKey: self.keyPrefix + key)
                callback(result, nil)

            case "setItem":
                guard args.count >= 2,
                      let key = args[0] as? String,
                      let val = args[1] as? String else {
                    callback(nil, "setItem: missing key or value")
                    return
                }
                self.defaults.set(val, forKey: self.keyPrefix + key)
                callback(nil, nil)

            case "removeItem":
                guard let key = args.first as? String else {
                    callback(nil, "removeItem: missing key")
                    return
                }
                self.defaults.removeObject(forKey: self.keyPrefix + key)
                callback(nil, nil)

            case "getAllKeys":
                let allKeys = self.defaults.dictionaryRepresentation().keys
                    .filter { $0.hasPrefix(self.keyPrefix) }
                    .map { String($0.dropFirst(self.keyPrefix.count)) }
                callback(allKeys, nil)

            case "clear":
                let allKeys = self.defaults.dictionaryRepresentation().keys
                    .filter { $0.hasPrefix(self.keyPrefix) }
                for key in allKeys {
                    self.defaults.removeObject(forKey: key)
                }
                callback(nil, nil)

            default:
                callback(nil, "AsyncStorageModule: Unknown method '\(method)'")
            }
        }
    }
}
